File: HostApp/FULLMONI-WIDE-Terminal/Models/ReleaseManifest.cs
using System.Text.Json.Serialization;

namespace FullmoniTerminal.Models;

/// <summary>
/// リリースマニフェスト全体
/// </summary>
public class ReleaseManifest
{
    [JsonPropertyName("schemaVersion")]
    public int SchemaVersion { get; set; }

    [JsonPropertyName("version")]
    public string Version { get; set; } = string.Empty;

    [JsonPropertyName("releaseDate")]
    public string ReleaseDate { get; set; } = string.Empty;

    [JsonPropertyName("releaseNotes")]
    public string? ReleaseNotes { get; set; }

    [JsonPropertyName("firmware")]
    public FirmwareInfo? Firmware { get; set; }

    [JsonPropertyName("hostApps")]
    public HostAppsInfo? HostApps { get; set; }
}

/// <summary>
/// ファームウェア情報
/// </summary>
public class FirmwareInfo
{
    [JsonPropertyName("minimumBootloaderVersion")]
    public string MinimumBootloaderVersion { get; set; } = "1.0.0";

    [JsonPropertyName("variants")]
    public List<FirmwareVariant> Variants { get; set; } = new();
}

/// <summary>
/// ファームウェアバリアント（AppWizardリソース単位）
/// </summary>
public class FirmwareVariant
{
    [JsonPropertyName("id")]
    public string Id { get; set; } = string.Empty;

    [JsonPropertyName("name")]
    public string Name { get; set; } = string.Empty;

    [JsonPropertyName("description")]
    public string? Description { get; set; }

    [JsonPropertyName("file")]
    public string File { get; set; } = string.Empty;

    [JsonPropertyName("thumbnail")]
    public string? Thumbnail { get; set; }

    [JsonPropertyName("size")]
    public long Size { get; set; }

    [JsonPropertyName("sha256")]
    public string Sha256 { get; set; } = string.Empty;

    /// <summary>
    /// ダウンロード用の完全URL（実行時に設定）
    /// </summary>
    [JsonIgnore]
    public string? DownloadUrl { get; set; }

    /// <summary>
    /// サムネイル用の完全URL（実行時に設定）
    /// </summary>
    [JsonIgnore]
    public string? ThumbnailUrl { get; set; }
}

/// <summary>
/// ホストアプリケーション情報
/// </summary>
public class HostAppsInfo
{
    [JsonPropertyName("windows")]
    public HostAppInfo? Windows { get; set; }

    [JsonPropertyName("android")]
    public HostAppInfo? Android { get; set; }
}

/// <summary>
/// 個別のホストアプリ情報
/// </summary>
public class HostAppInfo
{
    [JsonPropertyName("version")]
    public string Version { get; set; } = string.Empty;

    [JsonPropertyName("file")]
    public string File { get; set; } = string.Empty;

    [JsonPropertyName("size")]
    public long Size { get; set; }

    [JsonPropertyName("minSdk")]
    public int? MinSdk { get; set; }

    [JsonPropertyName("sha256")]
    public string Sha256 { get; set; } = string.Empty;
}

File: HostApp/FULLMONI-WIDE-Android/app/src/main/java/com/fullmoni/terminal/model/ReleaseManifest.kt
package com.fullmoni.terminal.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * リリースマニフェスト全体
 * GitHub Releases の release-manifest.json をパースするためのデータクラス
 */
@Serializable
data class ReleaseManifest(
    @SerialName("schemaVersion")
    val schemaVersion: Int = 1,
    
    @SerialName("version")
    val version: String = "",
    
    @SerialName("releaseDate")
    val releaseDate: String = "",
    
    @SerialName("releaseNotes")
    val releaseNotes: String? = null,
    
    @SerialName("firmware")
    val firmware: FirmwareInfo? = null,
    
    @SerialName("hostApps")
    val hostApps: HostAppsInfo? = null
)

/**
 * ファームウェア情報
 */
@Serializable
data class FirmwareInfo(
    @SerialName("minimumBootloaderVersion")
    val minimumBootloaderVersion: String = "",
    
    @SerialName("variants")
    val variants: List<FirmwareVariant> = emptyList()
)

/**
 * ファームウェアバリアント（AppWizardリソース単位）
 */
@Serializable
data class FirmwareVariant(
    @SerialName("id")
    val id: String = "",
    
    @SerialName("name")
    val name: String = "",
    
    @SerialName("description")
    val description: String? = null,
    
    @SerialName("file")
    val file: String = "",
    
    @SerialName("thumbnail")
    val thumbnail: String? = null,
    
    @SerialName("size")
    val size: Long = 0,
    
    @SerialName("sha256")
    val sha256: String = ""
) {
    /**
     * ダウンロード用の完全URL（実行時に設定）
     */
    @kotlinx.serialization.Transient
    var downloadUrl: String = ""
    
    /**
     * サムネイルURL（実行時に設定）
     */
    @kotlinx.serialization.Transient
    var thumbnailUrl: String = ""
    
    /**
     * サイズを読みやすい形式で表示
     */
    val sizeFormatted: String
        get() {
            return when {
                size >= 1024 * 1024 -> String.format("%.2f MB", size / (1024.0 * 1024.0))
                size >= 1024 -> String.format("%.1f KB", size / 1024.0)
                else -> "$size bytes"
            }
        }
}

/**
 * ホストアプリ情報
 */
@Serializable
data class HostAppsInfo(
    @SerialName("android")
    val android: HostAppInfo? = null,
    
    @SerialName("windows")
    val windows: HostAppInfo? = null
)

/**
 * 個別ホストアプリ情報
 */
@Serializable
data class HostAppInfo(
    @SerialName("version")
    val version: String = "",
    
    @SerialName("file")
    val file: String? = null,
    
    @SerialName("size")
    val size: Long = 0,
    
    @SerialName("sha256")
    val sha256: String = ""
)
